src/decode-l2tp.c

```diff
@@ -0,0 +1,549 @@
+    }
+
+    DetectPortCleanupList(NULL, head);
+}
+
+void DecodeL2TPConfig(void)
+{
+    int enabled = 0;
+    if (SCConfGetBool("decoder.l2tp.enabled", &enabled) == 1) {
+        if (enabled) {
+            g_l2tp_enabled = true;
+        } else {
+            g_l2tp_enabled = false;
+        }
+    }
+
+    int strict = 0;
+    if (SCConfGetBool("decoder.l2tp.strict", &strict) == 1) {
+        if (strict) {
+            g_l2tp_strict = true;
+        } else {
+            g_l2tp_strict = false;
+        }
+    }
+
+    if (g_l2tp_enabled) {
+        SCConfNode *node = SCConfGetNode("decoder.l2tp.ports");
+        if (node && node->val) {
+            DecodeL2TPConfigPorts(node->val);
+        } else {
+            DecodeL2TPConfigPorts(L2TP_DEFAULT_PORT_S);
+        }
+    }
+}
+
+/** \param pkt payload data directly above UDP or IP header
+ *  \param len length in bytes of pkt
+ */
+int DecodeL2TP(ThreadVars *tv, DecodeThreadVars *dtv, Packet *p, const uint8_t *pkt, uint32_t len)
+{
+    if (unlikely(pkt == NULL))
+        return TM_ECODE_FAILED;
+
+    /* General L2TP packet validation */
+    if (unlikely(!g_l2tp_enabled))
+        return TM_ECODE_FAILED;
+
+    if (unlikely(len < L2TP_MIN_HEADER_LEN)) {
+        ENGINE_SET_INVALID_EVENT(p, L2TP_PKT_TOO_SMALL);
+        return TM_ECODE_FAILED;
+    }
+
+    if (!PacketIncreaseCheckLayers(p)) {
+        return TM_ECODE_FAILED;
+    }
+
+    if (p->proto == IPPROTO_UDP) {
+        const L2TPoverUDPDataHdr *l2tp_hdr = (const L2TPoverUDPDataHdr *)pkt;
+        /* L2F / L2TPv1 / L2TPv2 decoding is not implemented, just return OK */
+        if (unlikely(l2tp_hdr->version == 1 || l2tp_hdr->version == 2)) {
+            StatsIncr(tv, dtv->counter_l2tp);
+            StatsIncr(tv, dtv->counter_l2tp_unsupported);
+            return TM_ECODE_OK;
+        } else if (unlikely(l2tp_hdr->version != 3)) {
+            SCLogDebug("L2TP Invalid. Type: %u Version: %u Reserved: %u", l2tp_hdr->type,
+                    l2tp_hdr->version, l2tp_hdr->reserved);
+            if (g_l2tp_strict) {
+                ENGINE_SET_INVALID_EVENT(p, L2TP_INVALID_VER);
+                return TM_ECODE_FAILED;
+            } else {
+                ENGINE_SET_EVENT(p, L2TP_INVALID_VER);
+                return TM_ECODE_OK;
+            }
+        } else if (unlikely(g_l2tp_strict && (l2tp_hdr->reserved != 0 || l2tp_hdr->type != 0))) {
+            return TM_ECODE_FAILED;
+        }
+        len -= sizeof(L2TPoverUDPDataHdr);
+        pkt += sizeof(L2TPoverUDPDataHdr);
+    } else if (unlikely(p->proto != IPPROTO_L2TP)) {
+        SCLogDebug("Invalid IP Protocol for L2TP (%u)", p->proto);
+        ENGINE_SET_INVALID_EVENT(p, L2TP_INVALID_IP_PROTO);
+        return TM_ECODE_FAILED;
+    }
+
+    SCLogDebug("L2TPv3 found. Protocol is: %u", p->proto);
+    StatsIncr(tv, dtv->counter_l2tp);
+
+#ifdef DEBUG
+    uint32_t session_id = (pkt[0] << 24) + (pkt[1] << 16) + (pkt[2] << 8) + pkt[3];
+#endif
+    len -= sizeof(uint32_t);
```

Comment:
I think we should have a `DEBUG_VALIDATE_BUG_ON(len < 4)` before the statements where we are subtracting from `len` so the fuzzers could find anything off.

Here for example, if we consume 4 bytes earlier in the if block, there are no guards left for `len`. So, for len = [4, 5, 6, 7], we can underflow. Perhaps a check should be added? Please lmk wdyt?